cpu/kinetis_common/periph/uart.c

```diff
@@ -177,46 +297,126 @@ static int init_base(uart_t uart, uint32_t baudrate)
     dev->CFIFO = UART_CFIFO_RXFLUSH_MASK | UART_CFIFO_TXFLUSH_MASK;
 #endif
 
-    /* enable transmitter and receiver */
-    dev->C2 |= UART_C2_TE_MASK | UART_C2_RE_MASK;
+    /* enable transmitter */
+    dev->C2 |= UART_C2_TE_MASK;
     return 0;
 }
 
 void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
-    KINETIS_UART *dev;
+    KINETIS_UART *dev = _uart_base_ptrs[uart];
 
-    switch (uart) {
-#if UART_0_EN
-        case UART_0:
-            dev = UART_0_DEV;
-            break;
-#endif
-#if UART_1_EN
-        case UART_1:
-            dev = UART_1_DEV;
-            break;
-#endif
-        default:
-            return;
+    if (_uart_dma_config[uart].channel < 0) {
+        for (size_t i = 0; i < len; i++) {
+            while (!(dev->S1 & UART_S1_TDRE_MASK)) {}
+            dev->D = data[i];
+        }
     }
+    else {
+        uint8_t channel = _uart_dma_config[uart].channel;
+        if (irq_is_in()) {
+            /* Put DMA on hold for the time being to avoid bus races */
+            BITBAND_REG8(dev->C2, UART_C2_TIE_SHIFT) = 0;
+            /* wait for any ongoing DMA transfer to complete */
+            while (DMA_DEV->TCD[channel].CSR & DMA_CSR_ACTIVE_MASK) {}
+
+            for (size_t i = 0; i < len; i++) {
+                while (!(dev->S1 & UART_S1_TDRE_MASK)) {}
+                dev->D = data[i];
+            }
+            /* Resume DMA transfers again */
+            BITBAND_REG8(dev->C2, UART_C2_TIE_SHIFT) = 1;
+        }
+        else {
+            mutex_lock(&_uart_write_mutex[uart]);
+            /* Select source address */
+            DMA_DEV->TCD[channel].SADDR = (uint32_t)data;
+            /* Source step size */
+            DMA_DEV->TCD[channel].SOFF = DMA_SOFF_SOFF(sizeof(*data));
+            /* Destination address */
+            DMA_DEV->TCD[channel].DADDR = (uint32_t)(&dev->D);
+            /* Destination step size */
+            DMA_DEV->TCD[channel].DOFF = DMA_DOFF_DOFF(0);
+            /* Set source and destination width to 1 byte */
+            DMA_DEV->TCD[channel].ATTR = DMA_ATTR_DSIZE(DMA_ACCESS_SIZE_8BIT) | DMA_ATTR_SSIZE(DMA_ACCESS_SIZE_8BIT);
+            /* Set minor loop count */
+            DMA_DEV->TCD[channel].NBYTES_MLOFFNO = DMA_NBYTES_MLOFFNO_NBYTES(1);
+            /* Set major loop count */
+            /* The DMA engine can only handle up to 2**15 major loop iterations */
+            assert(len <= (DMA_CITER_ELINKNO_CITER_MASK >> DMA_CITER_ELINKNO_CITER_SHIFT));
+            DMA_DEV->TCD[channel].CITER_ELINKNO = DMA_CITER_ELINKNO_CITER(len);
+            DMA_DEV->TCD[channel].BITER_ELINKNO = DMA_BITER_ELINKNO_BITER(len);
+            /* Set control register */
+            /* Disable DMA requests after major loop finishes, interrupt on end of major loop */
+            DMA_DEV->TCD[channel].CSR = DMA_CSR_DREQ_MASK | DMA_CSR_INTMAJOR_MASK;
```

Comment:
@gebart Can this DMA specific code (L333 ... L351) be moved to cpu/kinetis_common/periph/dma.c ?